operator/pkg/ciliumidentity/controller.go

```diff
@@ -0,0 +1,228 @@
+	cidQueue            workqueue.RateLimitingInterface
+	cidQueueRateLimiter *rate.Limiter
+	cidQueueQpsLimit    float64
+	cidQueueBurstLimit  int
+
+	podQueue workqueue.RateLimitingInterface
+	nsQueue  workqueue.RateLimitingInterface
+
+	// oldNSSecurityLabels is a map between namespace and it's security labels.
+	// It's used to track preivous state of labels, to detect when labels changed.
+	oldNSSecurityLabels map[string]labels.Labels
+
+	wp *workerpool.WorkerPool
+
+	cesEnabled            bool
+	googleMultiNICEnabled bool
+}
+
+type queueOperations interface {
+	enqueueCIDReconciliation(cidKey resource.Key)
+	enqueuePodReconciliation(podKey resource.Key)
+	enqueueNSReconciliation(nsKey resource.Key)
+}
+
+type queueItem struct {
+	key         resource.Key
+	enqueueTime time.Time
+}
+
+// registerController creates and initializes the CID controller
+func registerController(p params) {
+	if !p.Clientset.IsEnabled() || !p.SharedCfg.EnableOperatorManageCIDs {
+		return
+	}
+
+	cidController := &Controller{
+		logger:              p.Logger,
+		clientset:           p.Clientset,
+		namespaces:          p.Namespaces,
+		pods:                p.Pods,
+		ciliumIdentities:    p.CiliumIdentities,
+		ciliumEndpointSlice: p.CiliumEndpointSlice,
+		cidQueueQpsLimit:    p.Cfg.CIDQueueQPSLimit,
+		cidQueueBurstLimit:  p.Cfg.CIDQueueBurstLimit,
+		oldNSSecurityLabels: make(map[string]labels.Labels),
+		cesEnabled:          p.SharedCfg.EnableCiliumEndpointSlice,
+		metrics:             p.Metrics,
+	}
+
+	cidController.initializeQueues()
+
+	p.Lifecycle.Append(cidController)
+}
+
+func (c *Controller) Start(ctx hive.HookContext) error {
+	c.logger.Info("Starting Cilium Identity controller")
+	defer utilruntime.HandleCrash()
+
+	c.context, c.contextCancel = context.WithCancel(context.Background())
+	c.reconciler = newReconciler(
+		c.context,
+		c.logger,
+		c.clientset,
+		c.namespaces,
+		c.pods,
+		c.ciliumIdentities,
+		c.ciliumEndpointSlice,
+		c.cesEnabled,
+		c,
+	)
+
+	c.logger.Info("Starting CID controller reconciler.")
+
+	// The desired state needs to be calculated before the events are processed.
+	if err := c.reconciler.calcDesiredStateOnStartup(); err != nil {
+		return fmt.Errorf("CID controller failed to calculate the desired state: %v", err)
+	}
+
+	c.startWorkerPool()
+	c.startWorkQueues()
+
+	return nil
+}
+
+func (c *Controller) Stop(ctx hive.HookContext) error {
+	c.wp.Close()
+
+	c.cidQueue.ShutDown()
+	c.podQueue.ShutDown()
+	c.nsQueue.ShutDown()
+
+	c.contextCancel()
+	return nil
+}
+
+func (c *Controller) initializeQueues() {
+	c.initCIDQueue()
+	c.initPodQueue()
+	c.initNSQueue()
+}
+
+func (c *Controller) startWorkerPool() {
+	c.wp = workerpool.New(4)
+	c.wp.Submit("process-cilium-identity-events", c.processCiliumIdentityEvents)
+	c.wp.Submit("process-pod-events", c.processPodEvents)
+	c.wp.Submit("process-namespace-events", c.processNamespaceEvents)
+	c.wp.Submit("process-namespace-events", c.processCiliumEndpointSliceEvents)
```

Comment:
should be "process-cilium-endpointslice-events"?

Reply:
Done.